File: apps/mobile-app/android/app/src/main/java/net/aliasvault/app/autofill/utils/RustCredentialMatcher.kt
package net.aliasvault.app.autofill.utils

import android.util.Log
import net.aliasvault.app.vaultstore.models.Credential
import org.json.JSONArray
import org.json.JSONObject

/**
 * Wrapper for the Rust credential matcher using UniFFI bindings.
 */
object RustCredentialMatcher {
    private const val TAG = "RustCredentialMatcher"

    /**
     * Filter credentials based on app/website info using the Rust core credential matcher.
     *
     * @param credentials List of credentials to filter
     * @param searchText Search term (app package name, URL, or text)
     * @return Filtered list of credentials
     */
    fun filterCredentialsByAppInfo(
        credentials: List<Credential>,
        searchText: String,
    ): List<Credential> {
        // Early return for empty search
        if (searchText.isEmpty()) {
            return credentials
        }

        try {
            // Convert credentials to JSON format expected by Rust
            val rustCredentials = JSONArray()
            val credentialMap = mutableMapOf<String, Credential>()

            for (credential in credentials) {
                val idString = credential.id.toString()
                val credJson = JSONObject().apply {
                    put("Id", idString)
                    put("ServiceName", credential.service.name ?: JSONObject.NULL)
                    put("ServiceUrl", credential.service.url ?: JSONObject.NULL)
                    put("Username", credential.username ?: JSONObject.NULL)
                }
                rustCredentials.put(credJson)
                credentialMap[idString] = credential
            }

            // Prepare input JSON for Rust
            val input = JSONObject().apply {
                put("credentials", rustCredentials)
                put("current_url", searchText)
                put("page_title", "")
                put("matching_mode", "default")
            }

            // Call Rust via UniFFI
            val outputJson = uniffi.aliasvault_core.filterCredentialsJson(input.toString())

            // Parse output
            val output = JSONObject(outputJson)
            val matchedIds = output.getJSONArray("matched_ids")

            // If no matches found, return empty list
            if (matchedIds.length() == 0) {
                return emptyList()
            }

            // Convert matched IDs back to credentials, maintaining order
            val filtered = mutableListOf<Credential>()
            for (i in 0 until matchedIds.length()) {
                val id = matchedIds.getString(i)
                credentialMap[id]?.let { filtered.add(it) }
            }

            return filtered
        } catch (e: Exception) {
            Log.e(TAG, "Error filtering credentials with Rust matcher: ${e.message}", e)
            // Fallback to returning all credentials on error
            return credentials
        }
    }
}

File: apps/mobile-app/ios/VaultUI/Selection/Utils/RustCredentialMatcher.swift
import Foundation
import VaultModels
import RustCoreFramework

/// Wrapper for the Rust credential matcher using UniFFI bindings.
public class RustCredentialMatcher {

    /// Filter credentials based on search text using the Rust core credential matcher.
    ///
    /// - Parameters:
    ///   - credentials: List of credentials to filter
    ///   - searchText: Search term (app package name, URL, or text)
    /// - Returns: Filtered list of credentials
    public static func filterCredentials(_ credentials: [AutofillCredential], searchText: String) -> [AutofillCredential] {
        // Early return for empty search
        if searchText.isEmpty {
            return credentials
        }

        do {
            // Convert AutofillCredential to the format expected by Rust
            let rustCredentials = credentials.map { credential -> [String: Any?] in
                return [
                    "Id": credential.id.uuidString,
                    "ItemName": credential.serviceName,
                    "ItemUrls": credential.serviceUrls,
                    "Username": credential.username
                ]
            }

            // Prepare input JSON for Rust
            let input: [String: Any] = [
                "credentials": rustCredentials,
                "current_url": searchText,
                "page_title": "",
                "matching_mode": "default"
            ]

            let inputData = try JSONSerialization.data(withJSONObject: input, options: [])
            guard let inputJson = String(data: inputData, encoding: .utf8) else {
                print("[RustCredentialMatcher] Failed to create input JSON")
                return credentials
            }

            // Call Rust via UniFFI
            let outputJson = try filterCredentialsJson(inputJson: inputJson)

            // Parse output
            guard let outputData = outputJson.data(using: .utf8),
                  let output = try JSONSerialization.jsonObject(with: outputData) as? [String: Any],
                  let matchedIds = output["matched_ids"] as? [String] else {
                print("[RustCredentialMatcher] Failed to parse output JSON")
                return credentials
            }

            // If no matches found, return empty array
            if matchedIds.isEmpty {
                return []
            }

            // Convert matched IDs back to UUIDs
            let matchedUUIDs = matchedIds.compactMap { UUID(uuidString: $0) }

            // Filter and sort credentials by matched order
            let filtered = matchedUUIDs.compactMap { matchedId in
                credentials.first { $0.id == matchedId }
            }

            return filtered

        } catch {
            print("[RustCredentialMatcher] Error filtering credentials: \(error)")
            // Fallback to returning all credentials on error
            return credentials
        }
    }
}
